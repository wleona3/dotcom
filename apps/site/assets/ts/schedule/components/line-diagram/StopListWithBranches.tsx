import { isArray, last } from "lodash";
import React from "react";
import StopCard from "./StopCard";
import ExpandableBranch from "./ExpandableBranch";
import { CommonLineDiagramProps } from "./__line-diagram";
import { LineDiagramStop } from "../__schedule";
import { lineDiagramIndexes, isOnBranchLine } from "./line-diagram-helpers";

// build branched diagram structure
const buildBranchedLineDiagram = (
  lineDiagram: LineDiagramStop[]
): (LineDiagramStop | LineDiagramStop[])[] => {
  const branchedDiagram: (LineDiagramStop | LineDiagramStop[])[] = [];
  // determine if the diagram will collect branches ("inward")
  // or split into branches ("outward") going down the page
  // const direction = getTreeDirection(lineDiagram);

  // identify where in the list to branch out or in, e.g.
  // [stop, stop, stop, stop, stop, #, stop, stop]
  // # = on an "inward" direction tree, join the branch to the main line
  // [stop, stop, ^, stop, stop, stop, stop, stop]
  // ^ = on an "outward" direction tree, split the branch from the main line
  const toggledStopIndices = lineDiagramIndexes(
    lineDiagram,
    (stop: LineDiagramStop): boolean =>
      isOnBranchLine(stop) && last(stop.stop_data)!.type === "stop"
  );

  // iterate through the list, constructing branches as needed
  // branches will be grouped together, e.g.:
  // [stop, stop, [stop#, stop, stop], stop, stop]
  // # = branch terminus for a tree with "inward" direction
  // [stop, stop, [stop, stop, stop^], stop, stop]
  // ^ = branch terminus for a tree with "outward" direction
  lineDiagram.forEach((ld, i) => {
    if (toggledStopIndices.includes(i)) {
      // either place on branch or initialize new branch
      if (isArray(last(branchedDiagram))) {
        (last(branchedDiagram) as LineDiagramStop[]).push(ld);
      } else {
        branchedDiagram.push([ld]);
      }
    } else {
      branchedDiagram.push(ld);
    }
  });

  return branchedDiagram;
};

export const Branch = (props: CommonLineDiagramProps): React.ReactElement => {
  const { stops, handleStopClick, liveData } = props;
  return (
    <li>
      {stops.length > 2 ? (
        <ExpandableBranch key={`${stops[0].route_stop.id}-branch`} {...props} />
      ) : (
        <ol data-testid="branch">
          {stops.map(stop => (
            <StopCard
              key={stop.route_stop.id}
              stop={stop}
              onClick={handleStopClick}
              liveData={liveData?.[stop.route_stop.id]}
            />
          ))}
        </ol>
      )}
    </li>
  );
};

const StopListWithBranches = (
  props: CommonLineDiagramProps
): React.ReactElement => {
  const { stops, ...other } = props;
  const lineDiagramWithBranches = buildBranchedLineDiagram(stops);
  return (
    <ol>
      {lineDiagramWithBranches.map(stopOrStops => {
        if (isArray(stopOrStops)) {
          return (
            <Branch
              key={stopOrStops[0].route_stop.id}
              stops={stopOrStops}
              {...other}
            />
          );
        }

        // it's a stop, just render it
        return (
          <StopCard
            key={stopOrStops.route_stop.id}
            stop={stopOrStops}
            onClick={props.handleStopClick}
            liveData={props.liveData?.[stopOrStops.route_stop.id]}
          />
        );
      })}
    </ol>
  );
};

export default StopListWithBranches;
