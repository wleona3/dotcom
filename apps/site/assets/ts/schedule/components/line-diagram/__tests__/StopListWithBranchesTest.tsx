import React from "react";
import { screen } from "@testing-library/dom";
import { cloneDeep, merge } from "lodash";
import { RouteType } from "../../../../__v3api";
import { LineDiagramStop } from "../../__schedule";
import simpleLineDiagram from "./lineDiagramData/simple.json"; // not a full line diagram
import outwardLineDiagram from "./lineDiagramData/outward.json"; // not a full line diagram
import StopListWithBranches from "../StopListWithBranches";
import { mockedRenderWithStopPositionContext } from "../../../../__tests__/test-helpers";

const lineDiagram = (simpleLineDiagram as unknown) as LineDiagramStop[];
let lineDiagramBranchingOut = (outwardLineDiagram as unknown) as LineDiagramStop[];

const route = {
  type: 2 as RouteType,
  name: "route 1",
  long_name: "route 1 long name",
  color: "F00B42",
  id: "route-1",
  direction_names: {
    0: "Outbound",
    1: "Inbound"
  },
  direction_destinations: {
    0: "Begin",
    1: "End"
  },
  description: "key_bus_route",
  "custom_route?": false,
  header: "",
  alerts: []
};

lineDiagram.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

lineDiagramBranchingOut.forEach(({ route_stop }) => {
  route_stop.route = cloneDeep(route);
});

let lineDiagramBranchingIn = cloneDeep(lineDiagramBranchingOut).reverse();
const CRroute = merge(cloneDeep(route), { type: 2 as RouteType });
lineDiagramBranchingIn.forEach(({ route_stop }) => {
  route_stop.route = CRroute;
  if (route_stop["is_terminus?"]) {
    route_stop["is_beginning?"] = !route_stop["is_beginning?"];
  }
});

const handleStopClick = () => {};
const liveData = {};

test("StopListWithBranches renders and matches snapshot", () => {
  const { asFragment } = mockedRenderWithStopPositionContext(
    <StopListWithBranches
      stops={lineDiagramBranchingIn}
      handleStopClick={handleStopClick}
      liveData={liveData}
    />,
    lineDiagramBranchingIn
  );

  expect(asFragment()).toMatchSnapshot();
});
describe("StopListWithBranches", () => {
  let container: HTMLElement;
  beforeEach(() => {
    ({ container } = mockedRenderWithStopPositionContext(
      <StopListWithBranches
        stops={lineDiagramBranchingIn}
        handleStopClick={handleStopClick}
        liveData={liveData}
      />,
      lineDiagramBranchingIn
    ));
  });

  it("renders expandable branches as well as individual stops", () => {
    expect(
      container.querySelectorAll(".m-schedule-diagram__expander")
    ).toHaveLength(1);
    expect(
      container.querySelectorAll(".m-schedule-diagram__stop")
    ).toHaveLength(10);
  });

  it("renders short branches as expanded", () => {
    const branches = screen.queryAllByTestId("branch");
    expect(branches).toHaveLength(2);
    const shortBranch = branches[0];
    const longBranch = branches[1];
    expect(
      longBranch.querySelector(".m-schedule-diagram__expander")
    ).toBeDefined();
    expect(
      shortBranch.querySelector(".m-schedule-diagram__expander")
    ).toBeNull();
  });

  describe("where branching inward", () => {
    it.each`
      index | expectedBranchNaming
      ${0}  | ${"Destination Line"}
      ${1}  | ${null}
      ${2}  | ${null}
      ${3}  | ${null}
      ${4}  | ${"Twig Destination Line"}
      ${5}  | ${null}
      ${6}  | ${"Branch Destination Line"}
      ${7}  | ${null}
      ${8}  | ${null}
    `(
      "shows branch name $expectedBranchNaming at stop $index",
      ({ index, expectedBranchNaming }) => {
        const branchNameNode = container
          .querySelectorAll(".m-schedule-diagram__stop")
          [index].querySelector(".u-small-caps")!;

        if (expectedBranchNaming) {
          expect(branchNameNode.textContent).toEqual(expectedBranchNaming);
        } else {
          expect(branchNameNode).toBeNull();
        }
      }
    );
  });
});
