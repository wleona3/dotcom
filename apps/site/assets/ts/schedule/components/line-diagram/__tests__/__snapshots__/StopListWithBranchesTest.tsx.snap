// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopListWithBranches renders and matches snapshot 1`] = `
<DocumentFragment>
  <ol>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 36px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <div
          class="u-bold u-small-caps"
        >
          Destination Line
        </div>
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-destination"
            >
              <span>
                Destination
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 36px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-stopC"
            >
              <span>
                Stop C
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 36px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-stopB"
            >
              <span>
                Stop B
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 36px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-stopA"
            >
              <span>
                Stop A
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 52px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <div
          class="u-bold u-small-caps"
        >
          Twig Destination Line
        </div>
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-twig-destination"
            >
              <span>
                Twig Destination
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
    <li>
      <ol
        data-testid="branch"
      >
        <li
          class="m-schedule-diagram__stop"
          style="padding-left: 52px;"
        >
          <section
            class="m-schedule-diagram__content"
          >
            <header
              class="m-schedule-diagram__stop-heading"
            >
              <h4
                class="m-schedule-diagram__stop-link notranslate"
              >
                <a
                  href="/stops/tree-twig-stop"
                >
                  <span>
                    Twig Stop
                  </span>
                </a>
              </h4>
              <div
                class="m-schedule-diagram__features"
              />
            </header>
            <div
              class="m-schedule-diagram__stop-details"
            >
              <div
                class="m-schedule-diagram__connections"
              />
            </div>
            <footer
              class="m-schedule-diagram__footer"
            >
              <button
                class="btn btn-link"
                type="button"
              >
                View schedule
              </button>
            </footer>
          </section>
        </li>
      </ol>
    </li>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 68px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <div
          class="u-bold u-small-caps"
        >
          Branch Destination Line
        </div>
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-branch-destination"
            >
              <span>
                Branch Destination
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
    <li>
      <div
        class="m-schedule-diagram__expander"
      >
        <h3
          aria-controls="panel-TreeBranch"
          aria-expanded="false"
          class="c-expandable-block__header"
          id="header-TreeBranch"
          role="button"
          tabindex="0"
        >
          <div
            class="c-expandable-block__header-text"
          >
            <div
              class="m-schedule-diagram__expander-header"
            >
              <div
                style="width: 54px;"
              >
                <svg
                  class="line-diagram-svg__toggle commuter-rail"
                  height="100%"
                  width="54"
                >
                  <g
                    transform="translate(31, 1)"
                  >
                    <rect
                      height="42"
                      rx="8"
                      width="16"
                    />
                    <circle
                      cx="8"
                      cy="11"
                      r="4"
                    />
                    <circle
                      cx="8"
                      cy="21"
                      r="4"
                    />
                    <circle
                      cx="8"
                      cy="31"
                      r="4"
                    />
                  </g>
                </svg>
              </div>
              <button
                class="btn btn-link m-schedule-diagram__toggle"
                type="button"
              >
                3 stops
              </button>
            </div>
            <span
              class="fa c-expandable-block__header-caret"
            >
              ï„‡
            </span>
          </div>
        </h3>
        <noscript />
      </div>
    </li>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 36px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-stop1"
            >
              <span>
                Stop 1
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 36px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-stop0"
            >
              <span>
                Stop 0
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
    <li
      class="m-schedule-diagram__stop"
      style="padding-left: 36px;"
    >
      <section
        class="m-schedule-diagram__content"
      >
        <header
          class="m-schedule-diagram__stop-heading"
        >
          <h4
            class="m-schedule-diagram__stop-link notranslate"
          >
            <a
              href="/stops/tree-origin"
            >
              <span>
                Origin
              </span>
            </a>
          </h4>
          <div
            class="m-schedule-diagram__features"
          />
        </header>
        <div
          class="m-schedule-diagram__stop-details"
        >
          <div
            class="m-schedule-diagram__connections"
          />
        </div>
        <footer
          class="m-schedule-diagram__footer"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            View schedule
          </button>
        </footer>
      </section>
    </li>
  </ol>
</DocumentFragment>
`;
